jsonrpc/src/cancellation.ts

```diff
@@ -87,15 +89,58 @@ class MutableToken implements CancellationToken {
 	}
 }
 
-export class CancellationTokenSource {
+class FileBasedToken extends MutableToken {
+
+	constructor(private _cancellationName: string) {
+		super();
+	}
+
+	public cancel() {
+		if (!super.isCancellationRequested && createCancellationFile(this._cancellationName)) {
+			// change state only when writing cancellation file succeeded
+			// otherwise, ignore cancellation request
+			super.cancel();
+		}
+	}
+
+	get isCancellationRequested(): boolean {
+		if (super.isCancellationRequested) {
+			return super.isCancellationRequested;
+		}
+
+		if (this.pipeExists()) {
+			// the first time it encounters cancellation file, it will
+			// cancel itself and raise cancellation event.
+			// in this mode, cancel() might not be called explicitly by jsonrpc layer
+			this.cancel();
+		}
+
+		return super.isCancellationRequested;
+	}
+
+	private pipeExists(): boolean {
+		try {
+			fs.statSync(this._cancellationName);
+			return true;
+		}
+		catch (e) {
+			return false;
+		}
+	}
+}
+
+// internal usage only
+export class CancellationTokenSourceImpl {
```

Comment:
is there like "internal" in typescript to hide this type from outside?

Reply:
Everything that is not export is internal.

Reply:
so I want to use CancellationTokenSourceImpl in other file but only within json-rpc. is there a way to do that? or since it is not re-exported in main.ts, it is fine?

Reply:
Since it is not reexperted in main it is fine

I ran out of time today so a detailed review has to wait until tomorrow. One thing I already notices is that the cancelation support in JSON-RPC is pre request but your pipe looks global. I think we need to encode the request ID into the pipe name on the creation and checking side. Or do I miss something

Reply:
it is per request. I adds request id on each cancellation - https://github.com/microsoft/vscode-languageserver-node/pull/587/files#diff-6a132c34fec94e7b4e8e105aa67fdd15R461

Reply:
Now I see. I think this should be handled differently. We have an abstract class AbstrctCancellationTokenSource with two sub classes CancellationTokenSource and FileBaseCancellationTokenSource. FileBaseCancellationTokenSource will not be reexported.